Add view endpoint

Renny4Real opened the pull request:
Add endpoint that retrieves a single housing info record.
Have added additional (and hopefully helpful!) descriptions to commit messages to explain what changes were done per commit.

**Any suggestions, feedback, comments would be appreciated.**

Please let me know if something's missing or wrongly implemented.

Renny4Real commented on UHResidentInformationAPI.Tests/V1/Gateways/UHGatewayTests.cs:
```diff
@@ -1,51 +1,128 @@
+        public void GetResidentByIdReturnsPhoneContactDetails()
+        {
+            var databasePersonEntity = AddPersonRecordToDatabase();
+            var databasePhoneEntity = TestHelper.CreateDatabaseTelephoneNumberForPersonId(databasePersonEntity.ContactSid);
```
Need to check if this is what we use to link the person table and telephone contact tables.

Renny4Real commented on UHResidentInformationAPI/V1/Domain/ResidentInformation.cs:
```diff
@@ -21,10 +21,8 @@ public class ResidentInformation
     {
+        public string PropertyRef { get; set; }
         public string AddressLine1 { get; set; }
-        public string AddressLine2 { get; set; }
```
Removed these because we don't have multiple address columns in the property table.

Renny4Real committed:
Suprress analysis error: Specify IFormatProvider

Renny4Real committed:
Rename test database and add test DB dockerfile

Renny4Real committed:
Edit infrastructure classes to match schema descriptions

- This adds the `StringLength` descriptor to columns that are described as `char` in the schema as these string type are fixed in length.
- This adds methods fro columns that are NOT NULL in each table, so that fake data can be created for these during the tests.
- This renames the column & table headings to be lower case as it was found that they are created in lowercase in the database and are case sensitive when interacting with/querying them.
- This add the `Range` descriptor to describe a numeric column with data entries having a total number of 2 digits.
- This changes the values for the max length to match schema description.

Renny4Real committed:
Add gateway tests for getting single housing record by id

This adds tests to check that the response from the gateway contains:
- details about the person
- details about their address and UPRN
- details about their email addresses
- details about their phone numbers

Renny4Real committed:
Add tests describing use case for retrieving a single record using the gateway
Describes:
- Response object expected when gateway method returns with a value
- Exception thrown when gateway method returns NULL

Renny4Real committed:
Add tests for controller and end-to-end functionality.
This adds:
- Required services to the `Startup.cs`
- Route for view endpoint
- Test describing expected response from controller
- Test describing the expected response from interactions of controller, use case and gateway methods

elcorbs commented on UHResidentInformationAPI/V1/Factories/EntityFactory.cs:
```diff
@@ -10,13 +16,48 @@ public static ResidentInformation ToDomain(this Person databaseEntity)
+        public static Phone ToDomain(this TelephoneNumber databaseEntity)
+        {
+            var canParseType = Enum.TryParse<PhoneType>(databaseEntity.Type, out var type);
+            return canParseType ? new Phone
```
Will there always be a parsable phone type? Maybe we should be returning an object with `PhoneNumber` and `LastModified` and a `null` for `Type` if the type isn't parsable? I'm not sure though because I don't know what the data is like.

Renny4Real replied:
That's true. 
I could add some code to account for if the phone type is null. 
I just had a look at the schema again and the column for this is nullable.

Renny4Real replied:
Changes made in d235b521d286dc4ae1b83b279c4bbe2ec5eaa7e0

elcorbs commented on UHResidentInformationAPI/V1/Factories/ResponseFactory.cs:
```diff
@@ -0,0 +1,62 @@
+                LastName = domain.LastName,
+                DateOfBirth = domain.DateOfBirth,
+                NINumber = domain.NINumber,
+                PhoneNumber = domain.PhoneNumber.ToResponse(),
+                Email = domain.Email.ToResponse(),
+                Address = domain.Address.ToResponse()
```
Will these 3 fields always be there? If any of them is null this will throw a null value exception

Renny4Real replied:
Had a look at the schema: values for the email address and the phone number can't be null so these would probably always exist but the address and postcode could be null.

Will add code to account for this. Thanks 👍

Renny4Real replied:
Changes made in 6525ce07992b8187a2122491d6f4d5173ef5db40

elcorbs commented on UHResidentInformationAPI/V1/Factories/ResponseFactory.cs:
```diff
@@ -0,0 +1,62 @@
+            return emailAddresses.Select(email => new Email()
+            {
+                EmailAddress = email.EmailAddress,
+                LastModified = email.LastModified.ToString("O", CultureInfo.InvariantCulture)
```
Similarly, with the two Last Modified fields - will they always be populated, otherwise, it'll throw an exception

Renny4Real replied:
That makes sense.
The columns for this (`modDate`) have a `NOT NULL` description in the schema.
It might be safe to assume that these won't be null in the live DB.

elcorbs commented on UHResidentInformationAPI/V1/Gateways/UHGateway.cs:
```diff
@@ -21,7 +24,43 @@ public List<ResidentInformation> GetAllResidents(string houseReference, string r
+                _uHContext.EmailAddresses.Where(c => c.ContactID == databaseRecord.PersonNo).ToList();
+
+            var addressForPerson =
+                _uHContext.Addresses.FirstOrDefault(a => a.HouseRef == databaseRecord.HouseRef);
```
Should this be ordered by something before we pick the first record? Maybe so that we return the address which has been most recently modified?

Renny4Real replied:
That's true.
On hindsight, used this method based on the IDE suggestion when I attempted to use `where` to query it.
Could look into making sure the most up to date address entry is used. Maybe ordering it by a column that tracks the last modified date.

Renny4Real replied:
Changes made in 1a556eb

elcorbs reviewed (approved):
A few little comments but it looks good :)

Renny4Real committed:
Account for null address object from the gateway response.

- Address1 & postcode which make up the Address response are both nullable columns in the schema.
- This adds a statement in the ToResponse method to return an empty address response object when both the address fields are NULL.

Renny4Real committed:
Change Phone Type Enum to nullable type

The value of the phone type can be NULL in some cases:
- This changes the Phone Type Enum to be a nullable type.
- Adds a test that considers a phone object without a phone type being returned by the gateway.

louiseswift commented on database.env:
```diff
@@ -1,3 +1,3 @@
 POSTGRES_PASSWORD=mypassword
-POSTGRES_DB=entitycorex
\ No newline at end of file
+POSTGRES_DB=housinginfo-test
```
Although we don't really yet use a local development postgres service while working on these projects, technically the same `database.env` file contents here are used during setup of both the test and dev databases. (This file is referenced in both services in `docker-compose.yml`.)

This means that we might not want to append `-test` to the database name - or alternatively, in future, we might want to use a different file for the dev database's setup. In which case, no changes to make here now, and perhaps instead an update to the base-api for us to consider in future.

louisebolden reviewed (approved):
This is one of the most well-considered, organised and documented PRs I've ever seen. Very nice work!

Renny4Real committed:
Order addresses by the latest date before retrieving the first value

Renny4Real committed:
Edit test database infrastructure

From investigations done on the UH datasource, this commit:
 - adds primary key contraints for all the tables in the schema
 - adds infrastructure files for the contact link and tenancy agreement tables.
 - edits previous infrastructure files.

Renny4Real committed:
Change method of querying for phone details

From investigating the UH datasource it was found that:
 - querying for any specific contact detail e.g phone required a `contactNo` value which needs to be obtained from the `cccontactlink` table.
 - getting a specific `contactNo` value needs a `tag_ref` value from the `tenagree` table.
 - getting a specific `tag_ref` value needs the `house_ref` value from the `property` table.

This commit changes the previous way of querying for phone details from using `personNo` to using an obtained `contactNo` from the `cccontactlink` table.

Renny4Real committed:
Change method of querying for email details

From investigating the UH datasource it was found that:
     - querying for any specific contact detail e.g email required a `contactNo` value which needs to be obtained from the `cccontactlink` table.
     - getting a specific `contactNo` value needs a `tag_ref` value from the `tenagree` table.
     - getting a specific `tag_ref` value needs the `house_ref` value from the `property` table.

This commit changes the previous way of querying for email details from using `personNo` to using an obtained `contactNo` from the `cccontactlink` table.

Renny4Real commented:
From investigations done on the UH data source, it was found that:
     - To get a specific contact detail e.g email or phone, this required a `contactNo` value obtained from the `cccontactlink` table.
     - To get the`contactNo` value from `cccontactlink`, you needed a `tag_ref` value from the `tenagree` table.
     - To get a`tag_ref` value, you needed a `house_ref` value from the `property` table.

Have added changes to the previous way of querying for contact details:
i.e. This would instead use an obtained `contactNo` value from the `cccontactlink` table and not `personNo` from the `member` table.

Renny4Real committed:
Tidy up gateway method and tests

LBHMGeorgieva commented on UHResidentInformationAPI/V1/Gateways/UHGateway.cs:
```diff
@@ -21,7 +24,52 @@ public List<ResidentInformation> GetAllResidents(string houseReference, string r
+        private int? GetContactNoFromContactLink(string houseReference)
+        {
+            var tagReference = _uHContext.TenancyAgreements.FirstOrDefault(ta => ta.HouseRef == houseReference)?.TagRef;
+            var contactLinkUsingTagReference = _uHContext.ContactLinks.FirstOrDefault(co => co.TagRef == tagReference)?.ContactID;
```
This might return more than one ContactNo in case of joint tenancy. I know you have the first or default, but if we want the contact details for a specific person within a household, you need to also query by “Key2”, which is a column referring to the “person_no” column in the members table.

Renny4Real replied:
Changes added in ca97063

LBHMGeorgieva commented on UHResidentInformationAPI/V1/Infrastructure/ContactLink.cs:
```diff
@@ -0,0 +1,29 @@
+        [Key]
+        public int ContactID { get; set; }
+
+        [Column("key1")]
```
Key2 will be required, which corresponds to person_no from members table

Renny4Real replied:
Changes added in ca97063

LBHMGeorgieva commented on UHResidentInformationAPI.Tests/V1/Helper/TestHelper.cs:
```diff
@@ -0,0 +1,88 @@
+            return ft;
+        }
+
+        public static ContactLink CreateDatabaseContactLinkForPerson(string tagRef)
```
This will need to be amended to reflect the inclusion of person_no

Renny4Real replied:
Changes added in ca97063

MirelaGeorgiou reviewed (changes requested):
Looks good :) Please look at my comments regarding including person_no as it is crucial towards being able to retrieve contact info for more than one person in a household

LBHMKeyworth commented on UHResidentInformationAPI/V1/Gateways/UHGateway.cs:
```diff
@@ -21,7 +24,52 @@ public List<ResidentInformation> GetAllResidents(string houseReference, string r
+        private int? GetContactNoFromContactLink(string houseReference)
+        {
+            var tagReference = _uHContext.TenancyAgreements.FirstOrDefault(ta => ta.HouseRef == houseReference)?.TagRef;
+            var contactLinkUsingTagReference = _uHContext.ContactLinks.FirstOrDefault(co => co.TagRef == tagReference)?.ContactID;
```
As per the long slack chat I started it may be that "TagRef" isn't necessarily right here as it looks like "Key1" in the database sometimes holds a tagRef ("012345/02") and sometimes just the household number ("012345")...looks like we need to do some investigation before nailing down what we call these?

Renny4Real committed:
Add person No as a parameter to obtain specific contact details from the gateway

LBHMKeyworth commented on UHResidentInformationAPI/V1/Infrastructure/ContactLink.cs:
```diff
@@ -0,0 +1,29 @@
+
+        [Column("key1")]
+        [StringLength(20)]
+        public string TagRef { get; set; }
```
As per above - remains to be seen if TagRef is the right thing to call this!
Discussions shall happen :)

makeyw reviewed (commented):
Just added comments around the theme we were all discussing on Slack about exactly WHAT "Key1" field in CCContactLinks table holds...

Renny4Real commented:
`key_2 `in the [`cccontactlink` table](https://github.com/LBHackney-IT/housing-resident-information-api/blob/ca97063e802abcfec68d25c0e4d3878d70651b1c/database/schema.sql#L42) (referring to `personNo`) has a data type of `char(10)` in the schema but `personNo` in [`member` table](https://github.com/LBHackney-IT/housing-resident-information-api/blob/ca97063e802abcfec68d25c0e4d3878d70651b1c/database/schema.sql#L78) is numeric(2,0).

It's been added as a string in the contact link infrastructure file and converted where needed.
